pkg/controller/awsmachine/awsmachine_controller.go

```diff
@@ -69,18 +103,283 @@ type ReconcileAWSMachine struct {
+
+	// Fetch the AWSMachine instance.
+	awsm := &infrav1.AWSMachine{}
+	err := r.Get(ctx, request.NamespacedName, awsm)
 	if err != nil {
-		if errors.IsNotFound(err) {
-			// Object not found, return.  Created objects are automatically garbage collected.
-			// For additional cleanup logic use finalizers.
+		if apierrors.IsNotFound(err) {
 			return reconcile.Result{}, nil
 		}
-		// Error reading the object - requeue the request.
 		return reconcile.Result{}, err
 	}
 
+	// Store Machine early state to allow patching.
+	patchFrom := client.MergeFrom(awsm.DeepCopy())
+
+	// If the Machine hasn't been deleted and doesn't have a finalizer, add one.
+	if awsm.ObjectMeta.DeletionTimestamp.IsZero() {
+		if !util.Contains(awsm.Finalizers, clusterv1.MachineFinalizer) {
+			awsm.Finalizers = append(awsm.ObjectMeta.Finalizers, clusterv1.MachineFinalizer)
+			if err := r.Client.Patch(ctx, awsm, patchFrom); err != nil {
+				return reconcile.Result{}, errors.Wrapf(err, "failed to add finalizer to AWSMachine %q/%q", awsm.Namespace, awsm.Name)
+			}
+			// Since adding the finalizer updates the object return to avoid later update issues
+			return reconcile.Result{Requeue: true}, nil
+		}
+	}
+
+	// Fetch the Machine.
+	m, err := r.getMachineOwner(ctx, awsm.ObjectMeta)
+	if err != nil {
+		return reconcile.Result{}, err
+	} else if m == nil {
+		klog.Infof("Waiting for Machine Controller to set OwnerRef on AWSMachine %q/%q", awsm.Namespace, awsm.Name)
+		return reconcile.Result{RequeueAfter: 10 * time.Second}, nil
+	}
+
+	// Make sure bootstrap data is available and populated.
+	if m.Spec.Bootstrap.Data == nil || *m.Spec.Bootstrap.Data == "" {
+		klog.Infof("Waiting for bootstrap data to be available on AWSMachine %q/%q", awsm.Namespace, awsm.Name)
+		return reconcile.Result{RequeueAfter: 10 * time.Second}, nil
+	}
+
+	scope, err := actuators.NewMachineScope(actuators.MachineScopeParams{
+		ProviderMachine: awsm,
+		Machine:         m,
+		Client:          r.Client,
+	})
+	if err != nil {
+		return reconcile.Result{}, errors.Errorf("failed to create scope: %+v", err)
+	}
+	defer scope.Close()
+
 	return reconcile.Result{}, nil
 }
+
+func (r *ReconcileAWSMachine) reconcile(ctx context.Context, scope *actuators.MachineScope) error {
+	exist, err := r.exists(scope)
+	if err != nil {
+		klog.Errorf("Failed to check if Machine %q infrastructure exists: %v", scope.Name(), err)
+		return err
+	}
+
+	// Reconcile ProviderID.
+	if pid := scope.GetProviderID(); pid == nil || *pid == "" {
+		scope.SetProviderID(fmt.Sprintf("aws:////%s", *scope.GetInstanceID()))
+	}
+
+	if exist {
+		scope.Info("Reconciling Machine triggers idempotent update")
+		return r.update(scope)
+	}
+
+	// Machine resource created. Machine does not yet exist.
+	scope.Info("Reconciling Machine triggers idempotent create")
+	return r.create(scope)
+}
+
+// create creates a machine and is invoked by the machine controller.
+func (r *ReconcileAWSMachine) create(scope *actuators.MachineScope) error {
```

Comment:
Since this is eventually calling `CreateOrGetMachine` should this be named `createOrGet` instead?

Reply:
I don't think so, I copied the old actuator code. But this `create` call only happens if the `exist` call doesn't succeed

Reply:
We probably need to clean this up quite a bit down the line, no need to do additional queries against the AWS api if we've already verified the existence.

Reply:
Completely agree